x/reward/keeper/query_server.go

```diff
@@ -0,0 +1,202 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/types/query"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/provenance-io/provenance/x/reward/types"
+)
+
+var _ types.QueryServer = Keeper{}
+
+func (k Keeper) RewardPrograms(ctx context.Context, req *types.QueryRewardProgramsRequest) (*types.QueryRewardProgramsResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	var err error
+
+	rewardProgramStates := []types.RewardProgram_State{}
+	switch req.QueryType {
+	case types.QueryRewardProgramsRequest_QUERY_TYPE_PENDING:
+		rewardProgramStates = []types.RewardProgram_State{types.RewardProgram_STATE_PENDING}
+	case types.QueryRewardProgramsRequest_QUERY_TYPE_ACTIVE:
+		rewardProgramStates = []types.RewardProgram_State{types.RewardProgram_STATE_STARTED}
+	case types.QueryRewardProgramsRequest_QUERY_TYPE_FINISHED:
+		rewardProgramStates = []types.RewardProgram_State{types.RewardProgram_STATE_FINISHED, types.RewardProgram_STATE_EXPIRED}
+	case types.QueryRewardProgramsRequest_QUERY_TYPE_OUTSTANDING:
+		rewardProgramStates = []types.RewardProgram_State{types.RewardProgram_STATE_PENDING, types.RewardProgram_STATE_STARTED}
+	}
+
+	response := types.QueryRewardProgramsResponse{}
+	kvStore := sdkCtx.KVStore(k.storeKey)
+	prefixStore := prefix.NewStore(kvStore, types.RewardProgramKeyPrefix)
+	pageResponse, err := query.FilteredPaginate(prefixStore, req.Pagination, func(key []byte, value []byte, accumulate bool) (bool, error) {
+		if accumulate {
+			var rewardProgram types.RewardProgram
+			vErr := rewardProgram.Unmarshal(value)
+			switch {
+			case vErr == nil && len(rewardProgramStates) == 0:
+				response.RewardPrograms = append(response.RewardPrograms, rewardProgram)
+			case vErr == nil:
+				for _, state := range rewardProgramStates {
+					if rewardProgram.GetState() == state {
+						response.RewardPrograms = append(response.RewardPrograms, rewardProgram)
+						break
+					}
+				}
+			default:
+				return false, vErr
+			}
+		}
+		return true, nil
```

Comment:
This still has the problem of not properly counting things (same with the `FilteredPaginate` in `ClaimPeriodRewardDistributions`). It needs to return `true` or `false` the same, regardless of what `accumulate` is.

Lets say there are 100 entries, and we're doing limit 2. But we're filtering for `state` = `expired`. There are two expired entries, at indexes 1, 94, and 95.

Offset 0:
* `accumulate` will start out as `true`.
* Index 0 returns `false`, so it  isn't counted.
* Index 1 returns `true`, and is counted/added.
* Indexes 2-93 return `false`.
* Index 94, returns `true` and is counted/added.
* There have been two `true`s, so the limit has been reached, and `accumulate` is now `false`.
* If still counting, indexes 95, 96, 97, 98, and 99 all return true
* It has returned `true` 7 times, so that's how many total results it will claim to have.

Then, Offset = 2:
* `accumulate` will start out as `false`.
* Indexes 0, and 1 both return `true`, and are counted (but not added to the results).
* The count now equals the offset, so `accumulate` becomes `true`.
* Indexes 2-93 return `false`.
* Index 94 returns `true` and is counted/added.
* Index 95 returns `true` and is counted/added.
* There have been two `true`s, so the limit has been reached, and `accumulate` is `false` again.
* If still counting, indexes 96, 97, 98, and 99 all return `true`.
* It has returned `true` 8 times (2 from the start, 2 from accumulation, and 4 for after), and will report that there are now 8 results total.

Later, at Offset = 50:
* `accumulate` will start out as `false`.
* Indexes 0 to 49 all return `true`, and are counted (but not added to the results).
* The count now equals the offset, so `accumulate` becomes `true`.
* Indexes 50-93 return `false`.
* Index 94 returns `true` and is counted/added.
* Index 95 returns `true` and is counted/added.
* There have been two `true`s, so the limit has been reached, and `accumulate` is `false` again.
* If still counting, indexes 96, 97, 98, and 99 all return `true`.
* It has returned `true` 56 times (50 from the start, 2 from accumulation, and 4 for after), and will report that there are now 8 results total.
* It also returns the exact same two results that were returned with Offset = 2.

Reply:
great explanation! @SpicyLemon 

moved all accumulate to only where `append` is used